asm.js-style setjmp/longjmp handling for wasm

aheejin opened the pull request:
The following PRs are related:
https://github.com/WebAssembly/binaryen/pull/681
https://reviews.llvm.org/D23928

@dschuff @jpporto

aheejin commented:
Currently passes all setjmp/longjmp related tests in binaryen1, binaryen2, binaryen3 modes. In binaryen0 mode, only test_longjmp_stacked fails, and this is because of a wasm llvm backend bug. (This fails even without the function containing setjmp/longjmp.)

kripken commented on src/library.js:
```diff
@@ -2700,6 +2713,14 @@ LibraryManager.library = {
   },
+#if WASM_BACKEND == 1
+  // We need this wrapper from llvm wasm backend, but this is generated only
+  // in asm.js compilation, so.
```
i'm confused about the comment here?

aheejin replied:
I don't know why it is done that way, but in asm.js, the call sequence is
emscripten_longjmp__wrapper(env, value) -> emscripten_longjmp(env, value) -> _longjmp(env, value).

Coincidentally, wasm llvm pass needs this call sequence too, because in LLVM IR side we need emscripten_wrapper for two kinds of function signatures. The first one is 
emscripten_longjmp(jmp_buf, int), which is the same as the original longjmp. We need this because we have to replace calls to longjmp to this. But secondly we also need emscripten_longjmp(int, int), because we need to model this code in LLVM side:
https://github.com/kripken/emscripten-fastcomp/blob/incoming/lib/Target/JSBackend/CallHandlers.h#L229
Here we call emscripten_longjmp with its first argument int (the value loaded from __THREW__ global variable). In asm.js, this code is in JS side so it doesn't matter, but we need to do this in llvm, so we need to call a function with the signature (int, int).

That's the reason why we need two signatures of emscripten_longjmp in llvm pass. Currently, in the LLVM pass, I declare functions as
emscripten_longjmp__wrapper(jmp_buf, int)
emscripten_longjmp(int, int)
And emscrpten_longjmp__wrapper calls emscripten_longjmp, which calls _longjmp. emscripten_longjmp__wrapper's signature will be lowered to (int, int) eventually in the backend, but in the LLVM IR pass we need two separate signatures.

dschuff replied:
Where does `emscripten_longjmp_wrapper` come from in asm.js? I don't see any references to it in the emscripten or fastcomp codebases? So does this call to `_longjmp` call back into the asm/wasm module?
wrt the comment I think @azakai just meant that the sentence seems to end abruptly. And I guess it doesn't really say what it does, but maybe the full explanation doesn't belong in library.js anyway?

aheejin replied:
In asm.js, emscripten_longjmp__wrapper is generated here:
https://github.com/kripken/emscripten/blob/incoming/emscripten.py#L540
But actually I don't understand the purpose of this make_table and fix_item functions, and why asm.js generates this wrapper in the first place.

_longjmp exists in JS, so it does not call back into asm/wasm.
https://github.com/kripken/emscripten/blob/incoming/src/library.js#L2695

Yes, I don't know how much details I should add. I can add all the details I described here
https://github.com/kripken/emscripten/pull/4524#discussion_r76551772
as a comment, if you and Alon think it is better.

kripken replied:
Yes, in the asm.js path the `__wrapper` functions are created so that a JS library function can be called indirectly. In other words, we import `emscripten_longjmp`, and we can't put that in the function table (because asm.js doesn't like that, it wants to know the type), so we wrap it in an asm.js function.

aheejin replied:
I discussed with @dschuff about this and we think it might be better to do convert all calls to emscripten_longjmp__wrapper to emscripten_longjmp and delete this hack here. Actually it was really confusing because the wasm's usage of emscripten_longjmp__wrapper and that of asm.js were different and I named it the same.

kripken commented on src/library.js:
```diff
@@ -2634,9 +2634,22 @@ LibraryManager.library = {
   // setjmp.h
   // ==========================================================================
 
+  // asm.js-style setjmp/longjmp support for wasm
```
this first line suggests the code following is just for wasm. maybe it could be rephrased to explain how the two different paths work?

also, wasm compilation can be either wasm backend or asm2wasm, so wasm by itself implies both, but here we just mean the wasm backend, I think.

aheejin replied:
Done

aheejin commented:
About the new commit about setTempRet:
I forgot to commit this before. I don't really understand what the current code does, so I just deleted the existing code and added my code. This doesn't mean that I would like to commit this change. Please let me know what the existing code does and give some suggestions on how to modify the code to support both modes.

kripken commented on src/library.js:
```diff
@@ -2634,9 +2634,22 @@ LibraryManager.library = {
+#if WASM_BACKEND == 1
+  $setjmpId: 0,
+#endif
   saveSetjmp__asm: true,
```
please add a blank line here, so it's more clear the above wasm backend comment is not covering the rest of the code (which is not only for the wasm backend).

aheejin replied:
Done

kripken commented on src/parseTools.js:
```diff
@@ -1292,7 +1292,7 @@ function makeGetTempRet0() {
 
 function makeSetTempRet0(value) {
-  return RELOCATABLE ? "setTempRet0((" + value + ") | 0)" : ("tempRet0 = " + value);
+  return 'asm["setTempRet0"](' + value + ')';
```
why was `RELOCATABLE` removed? I guess because both code paths work using the `asm` object?

aheejin replied:
No, as I said I simply don't know what to do - I don't know what relocatable in emscripten means. (I know what that means in traditional sense, but don't know how it applies to emscripten.) And I don't know what the difference is between setTempRet0(value) and tempRet0 = value. That's the reason why I wrote here
https://github.com/kripken/emscripten/pull/4524#issuecomment-243477623
that it's not that I'd like to commit this but ask how things work here.

kripken replied:
Sorry, I missed your previous comment.

Relocatable is a mode kind of like dynamic linking. In it the asm.js backend basically replaces each read of a global from a constant address (normal mode) to `global_base` plus a constant offset. Then we can set the global base at runtime, in effect relocating it. We do the same for function pointers too. This allows a simple (but not very fast) form of dynamic linking in asm.js.

I believe this particular line of code needs to check if we are relocatable or not because an additional thing in relocatable mode, which is to not rely on asm.js globals - since we may link multiple relocatable asm.js modules. So we move e.g. `tempRet0` (a global we use to return values larger than JS can fit in a number) from the asm.js module into the outside, where we can only access it with a function call (and then multiple modules can access it). That makes e.g. returning an i64 between relocatable modules work.

In wasm we don't care about relocatable mode for now, although I do intend to prototype it in the asm2wasm mode soon. So the simplest thing here might be to just add a code path for `WASM_BACKEND` here.

aheejin replied:
Done

kripken commented on src/parseTools.js:
```diff
@@ -1292,7 +1292,10 @@ function makeGetTempRet0() {
 
 function makeSetTempRet0(value) {
-  return RELOCATABLE ? "setTempRet0((" + value + ") | 0)" : ("tempRet0 = " + value);
+  if (WASM_BACKEND == 1)
```
`{`, `}` around if and else bodies please

kripken replied:
lgtm with that

aheejin replied:
Done
